File: hacking_with_python/CSTI.py
import requests
from bs4 import BeautifulSoup
from termcolor import colored

# Function to check if the target website is vulnerable to CSTI
def check_csti_vulnerability(url):
    try:
        # Send a GET request to the target website
        response = requests.get(url)

        # Parse HTML content using BeautifulSoup
        soup = BeautifulSoup(response.text, 'html.parser')

        # Check for potential indicators of CSTI vulnerability
        if "{{" in soup.text and "}}" in soup.text:
            return True
        else:
            return False
    except Exception as e:
        print(colored("Error occurred while checking vulnerability:", 'red'), e)
        return False

# Function to provide guidance on testing the CSTI vulnerability
def test_csti_vulnerability():
    print(colored("\nTesting for Client-Side Template Injection vulnerability...\n", 'blue'))
    print("1. Look for input fields where user-supplied data is rendered.")
    print("2. Inject template code like {{7*7}} and observe if it gets evaluated.")
    print("3. Check for any unexpected behavior or responses indicating successful injection.")
    print("\nFor example, try injecting {{7*7}} in a search box and see if the result is 49.\n")

# Main function
def main():
    print(colored("Welcome to Client-Side Template Injection Vulnerability Tester\n", 'green'))
    target_url = input(colored("Enter your target website URL: ", 'cyan'))

    print(colored("\nChecking if the target website is vulnerable...\n", 'blue'))
    if check_csti_vulnerability(target_url):
        print(colored("The target website is potentially vulnerable to Client-Side Template Injection.", 'yellow'))
        test_csti_vulnerability()
    else:
        print(colored("The target website is not vulnerable to Client-Side Template Injection.", 'green'))

if __name__ == "__main__":
    main()

File: hacking_with_ruby/CSTI.rb
require 'colorize'
require 'uri'
require 'net/http'

def test_csti_vulnerability(url)
  puts "Testing for Client-Side Template Injection vulnerability in #{url}...".yellow
  begin
    uri = URI.parse(url)
    response = Net::HTTP.get_response(uri)
    if response.code == "200"
      puts "Website is accessible...".green
      puts "Checking for possible CSTI vulnerability...".yellow
      injection_code = "{{7*7}}"
      test_url = "#{url}/search?q=#{injection_code}"
      test_response = Net::HTTP.get_response(URI.parse(test_url))
      if test_response.body.include?("49")
        puts "The website appears to be vulnerable to CSTI!".red
        puts "To confirm the vulnerability, try injecting the following payload:".yellow
        puts "Payload: {{7*7}}".light_blue
        puts "If the result is 49, then the website is vulnerable.".yellow
      else
        puts "The website does not seem to be vulnerable to CSTI.".green
      end
    else
      puts "Unable to access the website. Response code: #{response.code}".red
    end
  rescue => e
    puts "An error occurred: #{e.message}".red
  end
end

puts "Enter your target website URL: ".yellow
target_website = gets.chomp

test_csti_vulnerability(target_website)
